### PR TITLE
NCBI registry

### LINKED ISSUE #11: NCBI Registry
http://www.ncbi.nlm.nih.gov/genbank/collab/db_xref

### PR DESCRIPTION
**Pull request by scolby33**
Fixes #11

### CODE REVIEW

**Comment by cthoyt** [comment]
I made a couple random stylistic changes. There are a few things to do still:

1. switch the data structure that gets returned to be a `Dict[str, Dict[str,str]` where `provider` is one of the keys
2. parse `cells[4]` to extract an example identifier for each source
3. parse `cells[0]` differently to get the URL href to the website for the prefix

**Comment by cthoyt** [comment]
you gotta parse the ID out of the example

Screen Shot 2021-03-14 at 00 23 16

File: src/bioregistry/align/ncbi.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+from bioregistry.align.utils import Aligner
+from bioregistry.external.ncbi import get_ncbi
+
+__all__ = ["NcbiAligner"]
+
+
+class NcbiAligner(Aligner):
+    key = "ncbi"
+    getter = get_ncbi
```
Comment (cthoyt) [inline src/bioregistry/align/ncbi.py:15]: if you add `curation_header = ('name',)` then it knows how to pretty print the curation table better

**Comment by cthoyt** [comment]
Result looks strange:
```
SRPDB                 Signal Recognition Particle Database                                                                                                                                               ['https://www.ncbi.nlm.nih.gov/genbank/collab/db_xref/ http://bio.lundberg.gu.se/dbs/SRPDB/SRPDB.html']
```

try running the align script for NCBI and looking at the curation table that gets printed

**Comment by scolby33** [comment]
> you gotta parse the ID out of the example
> 
> Screen Shot 2021-03-14 at 00 23 16

Oh, I thought it was the whole thing

**Comment by scolby33** [comment]
> Result looks strange:
> 
> ```
> SRPDB                 Signal Recognition Particle Database                                                                                                                                               ['https://www.ncbi.nlm.nih.gov/genbank/collab/db_xref/ http://bio.lundberg.gu.se/dbs/SRPDB/SRPDB.html']
> ```
> 
> try running the align script for NCBI and looking at the curation table that gets printed

Looks fine to me, there's just a really long description that screws up the rows

Screen Shot 2021-03-13 at 22 50 26

File: src/bioregistry/external/ncbi.py [hunk 2]
```diff
@@ -18,7 +18,7 @@
 
 def get_ncbi() -> Dict[str, Dict[str, str]]:
     """Get the NCBI data."""
-    path = BIOREGISTRY_MODULE.ensure(url=URL, name='ncbi.html')
```
Comment (cthoyt) [inline src/bioregistry/external/ncbi.py]: single quotes for lyfe

File: src/bioregistry/external/ncbi.py [hunk 3]
```diff
@@ -39,32 +39,38 @@ def get_ncbi() -> Dict[str, Dict[str, str]]:
...
-                item["generic_urls"] = [urlunsplit(url_parts)]
-            else:
-                item["generic_urls"] = link_href
-
-        example = cells[4].text.strip()
+            link_href = link.attrs["href"].strip()
+            if link_href:
+                url_parts = urlsplit(link_href)
+                if not url_parts.netloc:  # handle relative links
+                    if url_parts.path.startswith("/"):  # relative to site root
```
Comment (cthoyt) [inline src/bioregistry/external/ncbi.py:46]: flip if statement polarity and use elifs to make less indentation

Comment (cthoyt) [reply src/bioregistry/external/ncbi.py:46]: oh i see whats going on here. sad.

File: src/bioregistry/external/ncbi.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                        url_parts = (
+                            NCBI_URL_PARTS.scheme,
+                            NCBI_URL_PARTS.netloc,
+                            NCBI_URL_PARTS.path + url_parts.path,
+                            url_parts.query,
+                            url_parts.fragment,
+                        )
+                    item['generic_urls'] = [urlunsplit(url_parts)]
+                else:
+                    item['generic_urls'] = [link_href]
```
Comment (cthoyt) [inline src/bioregistry/external/ncbi.py:64]: shouldn't the URL just be a string? can you rename this field to "homepage"?

File: src/bioregistry/align/ncbi.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        """Stub for preparing NCBI xref data to be added to the bioregistry.
+
+        The NCBI getter does all the necessary processing, so this method is effectively an identity.
+
+        :param external_id: the key of the entry in the incoming data
+        :param external_entry: the actual entry data from the incoming data
+
+        :returns: the processed entry data (in this case, no actual processing is done)
+        """
+        return external_entry
```
Comment (cthoyt) [inline src/bioregistry/align/ncbi.py:34]: maybe you should just change this in the base class to be the default implementation

File: src/bioregistry/external/ncbi.py (hunk 4 above)
Comment (scolby33) [reply src/bioregistry/external/ncbi.py:64]: I copied the term and the list from GO, but that's fine too

File: src/bioregistry/external/ncbi.py (hunk 3 above)
Comment (scolby33) [reply src/bioregistry/external/ncbi.py]: Yeah, like 3 levels of validation :(

File: src/bioregistry/align/ncbi.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+        The NCBI getter does all the necessary processing, so use the superclass'
+        default implementation.
+
+        :param external_id: the key of the entry in the incoming data
+        :param external_entry: the actual entry data from the incoming data
+
+        :returns: the processed entry data
+        """
+        return super().prepare_external(external_id, external_entry)
```
Comment (cthoyt) [inline src/bioregistry/align/ncbi.py:35]: why bother overriding this? just don't provide any implementation at all

Comment (scolby33) [reply src/bioregistry/align/ncbi.py:35]: Because it's an abstractmethod, so I have to. I guess I could remove that annotation too if that makes sense.

Comment (cthoyt) [reply src/bioregistry/align/ncbi.py:35]: Yup. Remove the abstract tag

Comment (scolby33) [reply src/bioregistry/align/ncbi.py]: It's gone

**Review by cthoyt — APPROVED** [review]
Thanks @scolby33 !!! will merge when CI passes

**Pull request merged**